### PR TITLE
Upgrade to Stargate

### PR DESCRIPTION
**Pull request by alpe**
Upgrade to cosmos-sdk `6e569e125571`

Replaces https://github.com/CosmWasm/wasmd/pull/252

### CODE REVIEW

**Comment by ethanfrey** [comment]
The cosmos-sdk branching is a mess. Where is this commit `6e569e125571` from?

https://github.com/cosmos/cosmos-sdk/releases/tag/v0.40.0-rc1 is `68a23c4`, which would probably be the best to target now.

We cannot follow master (which may have non-0.40 breaking changes in it) and can only follow commits on `release/v0.40.x` branch (which luckily contains *both* `v0.40.0-rc0` and `v0.40.0-rc1`).

**Comment by ethanfrey** [comment]
Ow. "Refactor x/{gov, crisis} according to ADR 031". Yeah, this is Aaron's breaking change he is planning for 0.41.

*You can only trust tagged releases.* **Never trust master.** I am afraid you will have to redo a bit of this PR with that in mind. Hopefully not much :pray:

File: app/app.go [hunk 1]
```diff
@@ -401,42 +521,107 @@ func (app *WasmApp) EndBlocker(ctx sdk.Context, req abci.RequestEndBlock) abci.R
...
 
-// Codec returns the application's sealed codec.
-func (app *WasmApp) Codec() *codec.Codec {
-	return app.cdc
+// BlockedAddrs returns all the app's module account addresses that are not
+// allowed to receive external tokens.
+func (app *WasmApp) BlockedAddrs() map[string]bool {
+	blockedAddrs := make(map[string]bool)
+	for acc := range maccPerms {
+		blockedAddrs[authtypes.NewModuleAddress(acc).String()] = !allowedReceivingModAcc[acc]
```
Comment (ethanfrey) [inline app/app.go:551]: Nice one. I was about to say you needed to loop in both. But this is a clever way to get the diff in one loop

File: x/wasm/internal/keeper/handler_plugin.go [hunk 2]
```diff
@@ -121,12 +121,12 @@ func EncodeStakingMsg(sender sdk.AccAddress, msg *wasmTypes.StakingMsg) ([]sdk.M
 		if err != nil {
 			return nil, err
 		}
-		sdkMsg := staking.MsgDelegate{
-			DelegatorAddress: sender,
-			ValidatorAddress: validator,
+		sdkMsg := stakingtypes.MsgDelegate{
+			DelegatorAddress: sender.String(),
+			ValidatorAddress: validator.String(),
```
Comment (ethanfrey) [inline x/wasm/internal/keeper/handler_plugin.go:126]: We got the bech32 from the contract. On 116 we convert to binary, and here we convert back to bech32.

Why not delete 116-119 and replace this line with `ValidatorAddress: msg.Delegate.Validator`? Cleaner.

File: x/wasm/internal/keeper/handler_plugin.go [hunk 3]
```diff
@@ -99,12 +99,12 @@ func EncodeBankMsg(sender sdk.AccAddress, msg *wasmTypes.BankMsg) ([]sdk.Msg, er
 	if err != nil {
 		return nil, err
 	}
-	sdkMsg := bank.MsgSend{
-		FromAddress: fromAddr,
-		ToAddress:   toAddr,
+	sdkMsg := banktypes.MsgSend{
+		FromAddress: fromAddr.String(),
```
Comment (ethanfrey) [inline x/wasm/internal/keeper/handler_plugin.go:103]: Same here. And anywhere we see `sdk.AccAddressFromBech32()`. Just remove those calls and the `.String()` and use the address directly from the contract (the receiving module must parse and validate it)

File: x/wasm/internal/keeper/proposal_handler.go [hunk 4]
```diff
@@ -93,7 +93,15 @@ func handleMigrateProposal(ctx sdk.Context, k Keeper, p types.MigrateContractPro
 		sdk.NewAttribute(sdk.AttributeKeyModule, types.ModuleName),
 		sdk.NewAttribute(types.AttributeKeyContract, p.Contract.String()),
 	)
-	ctx.EventManager().EmitEvents(append(res.Events, ourEvent))
+	ctx.EventManager().EmitEvent(ourEvent)
+
+	for _, e := range res.Events {
```
Comment (ethanfrey) [inline x/wasm/internal/keeper/proposal_handler.go:98]: What is going on here>

res.Events.Attributes are in some abci format (`[]byte`?) which is different than the `sdk.Attribute`, so we grab them and convert them to strings for the event manager? (Which then converts them back to bytes to emit them likely)

File: x/wasm/internal/keeper/querier.go [hunk 5]
```diff
@@ -1,102 +1,168 @@
...
 }
 
-func queryContractListByCode(ctx sdk.Context, codeIDstr string, keeper Keeper) ([]byte, error) {
-	codeID, err := strconv.ParseUint(codeIDstr, 10, 64)
-	if err != nil {
+func (q grpcQuerier) SmartContractState(c context.Context, req *types.QuerySmartContractStateRequest) (*types.QuerySmartContractStateResponse, error) {
+	if err := sdk.VerifyAddressFormat(req.Address); err != nil {
+		return nil, err
+	}
+	ctx := sdk.UnwrapSDKContext(c).WithGasMeter(sdk.NewGasMeter(q.keeper.queryGasLimit))
```
Comment (ethanfrey) [inline x/wasm/internal/keeper/querier.go:107]: Nice to keep the gas meter here.

File: x/wasm/internal/keeper/querier.go [hunk 6]
```diff
@@ -1,102 +1,168 @@
...
+	return &types.QueryContractHistoryResponse{
+		Entries: rsp,
+	}, nil
 }
 
-func queryContractInfo(ctx sdk.Context, bech string, keeper Keeper) ([]byte, error) {
-	addr, err := sdk.AccAddressFromBech32(bech)
-	if err != nil {
-		return nil, sdkerrors.Wrap(sdkerrors.ErrInvalidAddress, err.Error())
+func (q grpcQuerier) ContractsByCode(c context.Context, req *types.QueryContractsByCodeRequest) (*types.QueryContractsByCodeResponse, error) {
```
Comment (ethanfrey) [inline x/wasm/internal/keeper/querier.go:55]: Wow, this whole file looks so much cleaner than the legacy one

File: x/wasm/internal/keeper/querier.go [hunk 7]
```diff
@@ -111,72 +177,22 @@ func queryContractListByCode(ctx sdk.Context, codeIDstr string, keeper Keeper) (
 	})
 	// and remove that info for the final json (yes, the json:"-" tag doesn't work)
```
Comment (ethanfrey) [inline x/wasm/internal/keeper/querier.go:178]: comment is outdated, right? there is no json marshaling anymore

File: x/wasm/internal/keeper/reflect_test.go [hunk 8]
```diff
@@ -72,14 +72,12 @@ func mustParse(t *testing.T, data []byte, res interface{}) {
 const MaskFeatures = "staking,mask"
 
 func TestMaskReflectContractSend(t *testing.T) {
-	tempDir, err := ioutil.TempDir("", "wasm")
-	require.NoError(t, err)
-	defer os.RemoveAll(tempDir)
-	ctx, keepers := CreateTestInput(t, false, tempDir, MaskFeatures, maskEncoders(MakeTestCodec()), nil)
-	accKeeper, keeper := keepers.AccountKeeper, keepers.WasmKeeper
+	cdc := MakeTestCodec()
```
Comment (ethanfrey) [inline x/wasm/internal/keeper/reflect_test.go:75]: Where do you create and destroy the tmpdir?

Maybe you create it in `CreateTestInput`, but you need to delete it after the tests of we just leave tons of orphan files.

File: x/wasm/internal/keeper/reflect_test.go [hunk 9]
```diff
@@ -451,10 +439,14 @@ type maskCustomMsg struct {
 	Raw   []byte `json:"raw,omitempty"`
 }
 
-// toMaskRawMsg encodes an sdk msg using amino json encoding.
+// toMaskRawMsg encodes an sdk msg using any type with json encoding.
```
Comment (ethanfrey) [inline x/wasm/internal/keeper/reflect_test.go:442]: woah, this is working again.
that is a nice feat. I would have probably removed it, but nice one

File: x/wasm/internal/keeper/test_common.go [hunk 10]
```diff
@@ -59,132 +113,149 @@ var TestingStakeParams = staking.Params{
...
-	keyAcc := sdk.NewKVStoreKey(auth.StoreKey)
-	keyStaking := sdk.NewKVStoreKey(staking.StoreKey)
-	keySupply := sdk.NewKVStoreKey(supply.StoreKey)
-	keyDistro := sdk.NewKVStoreKey(distribution.StoreKey)
-	keyParams := sdk.NewKVStoreKey(params.StoreKey)
-	tkeyParams := sdk.NewTransientStoreKey(params.TStoreKey)
+func CreateTestInput(t *testing.T, isCheckTx bool, supportedFeatures string, encoders *MessageEncoders, queriers *QueryPlugins) (sdk.Context, TestKeepers) {
+	tempDir, err := ioutil.TempDir("", "wasm")
+	require.NoError(t, err)
+	t.Cleanup(func() { os.RemoveAll(tempDir) })
```
Comment (ethanfrey) [inline x/wasm/internal/keeper/test_common.go:128]: Oh... I never knew about `t.Cleanup()`. That is a really nice test function! 

Ignore my comment above

File: x/wasm/internal/keeper/staking_test.go [hunk 11]
```diff
@@ -380,8 +373,8 @@ func TestReinvest(t *testing.T) {
 
 	// check some account values - the money is on neither account (cuz it is bonded)
 	// Note: why is this immediate? just test setup?
-	checkAccount(t, ctx, accKeeper, contractAddr, sdk.Coins{})
-	checkAccount(t, ctx, accKeeper, bob, funds)
+	checkAccount(t, ctx, accKeeper, bankKeeper, contractAddr, sdk.Coins{})
```
Comment (ethanfrey) [inline x/wasm/internal/keeper/staking_test.go:376]: Nice that so much of this code "just works" without changes. And the contracts as well.

File: x/wasm/internal/types/msg.proto [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+message MsgInstantiateContract {
+    option (gogoproto.goproto_getters) = false;
+
+    bytes sender = 1 [(gogoproto.casttype) = "github.com/cosmos/cosmos-sdk/types.AccAddress"];
+    // Admin is an optional address that can execute migrations
+    bytes admin = 2 [(gogoproto.casttype) = "github.com/cosmos/cosmos-sdk/types.AccAddress"];
+    uint64 code_id = 3 [(gogoproto.customname) = "CodeID"];
+    string label = 4;
+    bytes init_msg = 5 [(gogoproto.casttype) = "encoding/json.RawMessage"];
```
Comment (ethanfrey) [inline x/wasm/internal/types/msg.proto:34]: Clever. I was figuring we'd break this, but nice one.

**Review by ethanfrey — APPROVED** [review]
Looking good.

A bit of cleanup on the handler plugins, but generally ver nice job with this :1st_place_medal:

File: x/wasm/internal/keeper/proposal_handler.go (hunk 4 above)
Comment (alpe) [reply x/wasm/internal/keeper/proposal_handler.go:98]: Yes 🙈

**Review by ethanfrey — APPROVED** [review]
lgtm

**Comment by ethanfrey** [comment]
I would like to fix  codecov or remove it (it seems rather useless and non-deterministic). But that is for another PR. Let's get this in :rocket: :rocket:

**Pull request merged**